File: 백준/Silver/1010. 다리 놓기/다리 놓기.py
T = int(input())
def combi_to_dp(r,c) :
    if arr[r][c] != 0 :
        return arr[r][c]
    if r == c or c == 0 :
        return 1
    else :
        arr[r][c] = combi_to_dp(r-1,c-1) + combi_to_dp(r-1,c)
        return arr[r][c]
for t in range(T) :
    N, M = map(int,input().split())
    arr = [[0] * (N+1) for _ in range(M+1)]
    print(combi_to_dp(M,N))

File: 백준/Silver/1010. 다리 놓기/다리 놓기.java
import java.util.Scanner;

public class Main {
    static int[][] arr;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int T = scanner.nextInt();
        for (int t = 0; t < T; t++) {
            int N = scanner.nextInt();
            int M = scanner.nextInt();
            arr = new int[M + 1][N + 1];
            for (int i = 0; i <= M; i++) {
                for (int j = 0; j <= N; j++) {
                    arr[i][j] = 0;
                }
            }
            System.out.println(combiToDP(M, N));
        }
    }

    public static int combiToDP(int r, int c) {
        if (arr[r][c] != 0) {
            return arr[r][c];
        }
        if (r == c || c == 0) {
            return 1;
        } else {
            arr[r][c] = combiToDP(r - 1, c - 1) + combiToDP(r - 1, c);
            return arr[r][c];
        }
    }
}